Use SessionExpirationUtils for validate user and client sessions. Check client session is valid in TokenManager.

rmartinc opened the pull request:
Closes #24936

This PR checks the client session in the token manager to not manage a expired client session that generates negative values. Besides I have also unified the `AuthenticationManager.isSessionValid` to also use `SessionExpirationUtils` calculations. I'm sending a draft because of the last change. This is modifying some tests because previously the `isSessionValid` used just the remember-me values when the session was remember-me. Now it uses teh max value between the remember-me value and the normal sso value (as `SessionExpirationUtils` does). I think we should add a little upgrade note if we decide to continue this way. There are other two options: continue doing the calculation like before; change calculation in `SessionExpirationUtils` to use one or the other value. I prefer this way.

@mposolda @douglaspalmer Can you please test and review if you see something wrong?

rmartinc committed:
Use SessionExpirationUtils for validate user and client sessions
Check client session is valid in TokenManager
Closes #24936

Signed-off-by: rmartinc <rmartinc@redhat.com>

douglaspalmer reviewed (approved):
LGTM

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/RefreshTokenTest.java:
```diff
@@ -1356,6 +1349,168 @@ public void refreshTokenUserClientMaxLifespanGreaterThanSession() throws Excepti
     }
 
+    @Test
+    public void refreshTokenUserSessionMaxLifespanModifiedAfterTokenExchange() throws Exception {
```
```suggestion
    public void refreshTokenUserSessionMaxLifespanModifiedAfterTokenRefresh() throws Exception {
```

Minor nitpick to rename this method as it does not have anything to do with "token exchange" and hence is a little bit misleading.

Few other places in this class of similar naming.

rmartinc replied:
Done! I have changed all the methods...

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/RefreshTokenTest.java:
```diff
@@ -1356,6 +1349,168 @@ public void refreshTokenUserClientMaxLifespanGreaterThanSession() throws Excepti
+    }
+
+    @Test
+    public void refreshTokenClientSessionMaxLifespanModifiedAfterTokenExchange() throws Exception {
```
```suggestion
    public void refreshTokenClientSessionMaxLifespanModifiedAfterTokenRefresh() throws Exception {
```

rmartinc replied:
Done!

mposolda commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oauth/RefreshTokenTest.java:
```diff
@@ -1356,6 +1349,168 @@ public void refreshTokenUserClientMaxLifespanGreaterThanSession() throws Excepti
+    }
+
+    @Test
+    public void silentLoginClientSessionMaxLifespanModifiedAfterTokenExchange() throws Exception {
```
```suggestion
    public void silentLoginClientSessionMaxLifespanModifiedAfterTokenRefresh() throws Exception {
```

rmartinc replied:
Done!

mposolda commented on services/src/main/java/org/keycloak/protocol/oidc/TokenManager.java:
```diff
@@ -199,6 +199,12 @@ public TokenValidation validateToken(KeycloakSession session, UriInfo uriInfo, C
+        if (!AuthenticationManager.isClientSessionValid(realm, client, userSession, clientSession)) {
+            logger.debug("Client session is disabled");
+            userSession.removeAuthenticatedClientSessions(Collections.singletonList(client.getId()));
+            throw new OAuthErrorException(OAuthErrorException.INVALID_GRANT, "Client session disabled");
```
Nitpick: Isn't it "Client session expired" more accurate than "Client session disabled" ? If yes, similar probably applies also for debug message above.

rmartinc replied:
Yep, you are right, I have changed it to `Client session not active` to be consistent with the user session message.

mposolda reviewed (commented):
@rmartinc @douglaspalmer Thanks for the fix and review!

I've added some nitpick comments, which can be easily fixed (if you think they make sense) as the rebase in fixing conflict is needed anyway.

+1 for adding some migration documentation

rmartinc commented:
@mposolda Moving from draft as we all agreed to change the calculation like in the other parts. Requested changes added and rebased because there was a conflict.

rmartinc commented:
The new Persistent Session IT tests were failing. The reason seems to be that if the client session is removed and added in the same transaction the old session remains. I have changed to (instead of remove/add) do a restart similar to what we do in the user session. I'm adding @ahus1 to review because maybe I'm doing something wrong. :smile: 
All tests pass now in my branch.

ahus1 commented:
@rmartinc - can you please point me to where you did this change?

rmartinc commented:
> @rmartinc - can you please point me to where you did this change?

Yes, [here](https://github.com/keycloak/keycloak/pull/29013/files#diff-fd8b1c47da5ec7ee42d5450d3f7b73cdeef9c08af26b825c9245e3e6052c5cdfR578). In the previous version of the PR there was a remove/create of the client session. Now I created a `restartClientSession` that just cleans/adds the default client session values again. For that I added that method in [AuthenticatedClientSessionModel](https://github.com/keycloak/keycloak/pull/29013/files#diff-b4014a5e9c73d54dc98153e6fa64803e21ab0646a75566cf74ca9a02abbb18dbR77) and [AuthenticatedClientSessionAdapter](https://github.com/keycloak/keycloak/pull/29013/files#diff-c185363ae3c4ea873fe04e5057cd15688da7764fc306ca33157442ffa381c4d9R324)

mposolda reviewed (approved):
@rmartinc Thanks for the updates!

@ahus1 Could you please review or delegate to someone from the cross-dc team for the review?

ahus1 reviewed (approved):
Thank you for the PR, looks good to me.
